ext/io/fs.rs

```diff
@@ -0,0 +1,331 @@
+          metadata.$member()
+        }
+        #[cfg(not(unix))]
+        {
+          0
+        }
+      }};
+    }
+
+    #[inline(always)]
+    fn to_msec(maybe_time: Result<SystemTime, io::Error>) -> Option<u64> {
+      match maybe_time {
+        Ok(time) => Some(
+          time
+            .duration_since(UNIX_EPOCH)
+            .map(|t| t.as_millis() as u64)
+            .unwrap_or_else(|err| err.duration().as_millis() as u64),
+        ),
+        Err(_) => None,
+      }
+    }
+
+    Self {
+      is_file: metadata.is_file(),
+      is_directory: metadata.is_dir(),
+      is_symlink: metadata.file_type().is_symlink(),
+      size: metadata.len(),
+
+      mtime: to_msec(metadata.modified()),
+      atime: to_msec(metadata.accessed()),
+      birthtime: to_msec(metadata.created()),
+
+      dev: unix_or_zero!(dev),
+      ino: unix_or_zero!(ino),
+      mode: unix_or_zero!(mode),
+      nlink: unix_or_zero!(nlink),
+      uid: unix_or_zero!(uid),
+      gid: unix_or_zero!(gid),
+      rdev: unix_or_zero!(rdev),
+      blksize: unix_or_zero!(blksize),
+      blocks: unix_or_zero!(blocks),
+    }
+  }
+}
+
+#[async_trait::async_trait(?Send)]
+pub trait File {
+  fn read_sync(self: Rc<Self>, buf: &mut [u8]) -> FsResult<usize>;
+  fn write_sync(self: Rc<Self>, buf: &[u8]) -> FsResult<usize>;
+
+  fn write_all_sync(self: Rc<Self>, buf: &[u8]) -> FsResult<()>;
+  async fn write_all_async(self: Rc<Self>, buf: Vec<u8>) -> FsResult<()> {
+    self.write_all_byob(buf.into()).await
+  }
+
+  fn read_all_sync(self: Rc<Self>) -> FsResult<Vec<u8>>;
+  async fn read_all_async(self: Rc<Self>) -> FsResult<Vec<u8>>;
+
+  fn chmod_sync(self: Rc<Self>, pathmode: u32) -> FsResult<()>;
+  async fn chmod_async(self: Rc<Self>, mode: u32) -> FsResult<()>;
+
+  fn seek_sync(self: Rc<Self>, pos: io::SeekFrom) -> FsResult<u64>;
+  async fn seek_async(self: Rc<Self>, pos: io::SeekFrom) -> FsResult<u64>;
+
+  fn datasync_sync(self: Rc<Self>) -> FsResult<()>;
+  async fn datasync_async(self: Rc<Self>) -> FsResult<()>;
+
+  fn sync_sync(self: Rc<Self>) -> FsResult<()>;
+  async fn sync_async(self: Rc<Self>) -> FsResult<()>;
+
+  fn stat_sync(self: Rc<Self>) -> FsResult<FsStat>;
+  async fn stat_async(self: Rc<Self>) -> FsResult<FsStat>;
+
+  fn lock_sync(self: Rc<Self>, exclusive: bool) -> FsResult<()>;
+  async fn lock_async(self: Rc<Self>, exclusive: bool) -> FsResult<()>;
+
+  fn unlock_sync(self: Rc<Self>) -> FsResult<()>;
+  async fn unlock_async(self: Rc<Self>) -> FsResult<()>;
+
+  fn truncate_sync(self: Rc<Self>, len: u64) -> FsResult<()>;
+  async fn truncate_async(self: Rc<Self>, len: u64) -> FsResult<()>;
+
+  fn utime_sync(
+    self: Rc<Self>,
+    atime_secs: i64,
+    atime_nanos: u32,
+    mtime_secs: i64,
+    mtime_nanos: u32,
+  ) -> FsResult<()>;
+  async fn utime_async(
+    self: Rc<Self>,
+    atime_secs: i64,
+    atime_nanos: u32,
+    mtime_secs: i64,
+    mtime_nanos: u32,
+  ) -> FsResult<()>;
+
+  // deno_core
+  async fn read_byob(
```

Comment:
Add a `async fn read` also. Don't assume the implementation of `read` delegates to `read_byob`.